src/storage/src/source/mod.rs

```diff
@@ -801,72 +803,58 @@ where
-                                            .or_default()
-                                            .push(Ok(message))
-                                    }
+                                    let offset = message.offset;
+                                    untimestamped_messages
+                                        .entry(message.partition.clone())
+                                        .or_default()
+                                        .push((message, offset));
                                 }
                             }
                             // TODO: make errors definite
-                            Some(Err(e)) => pending_messages.push(Err(e)),
+                            Some(Err(e)) => non_definite_errors.push(e),
                             None => {},
                         }
                     }
                     // It's time to timestamp a batch
                     _ = timestamp_interval.tick() => {
-                        let mut max_offsets = HashMap::new();
-                        for message in pending_messages.iter().filter_map(|m| m.as_ref().ok()) {
-                            let entry = max_offsets.entry(message.partition.clone()).or_default();
-                            *entry = std::cmp::max(*entry, message.offset);
-                        }
-                        let (bindings, progress) = match timestamper.timestamp_offsets(&max_offsets).await {
-                            Ok((bindings, progress)) => (bindings, progress),
-                            Err(e) => {
-                                error!("Error timestamping offsets: {}", e);
-                                return;
+                        println!("reclocking stuff");
+                        let mut batch_upper = HashMap::new();
+                        for (pid, messages) in &untimestamped_messages {
+                            if let Some((_, offset)) = messages.last() {
+                                batch_upper.insert(pid.clone(), *offset + 1);
                             }
-                        };
-
-                        for msg in pending_messages.drain(..) {
-                            match msg{
-                                Ok(message) => {
-                                    let ts = bindings.get(&message.partition).expect("timestamper didn't return partition").0;
-                                    yield Event::Message(ts, Ok(message));
-                                },
-                                Err(e) => {
-                                    // TODO: make errors definite
-                                    yield Event::Message(0, Err(e));
-                                },
+                        }
+                        for (_, part_messages) in timestamper.reclock(&mut untimestamped_messages, &batch_upper).await {
+                            for (message, ts) in part_messages {
+                                println!("producing messages");
```

Comment:
Item No 3. Scalpel left in patient! The docker logs are very large due to this message, so random Buildkite issues are to be expected.

Reply:
fixed too